clang/lib/Parse/ParseStmt.cpp

```diff
@@ -2427,3 +2447,75 @@ bool Parser::ParseOpenCLUnrollHintAttribute(ParsedAttributes &Attrs) {
   }
   return true;
 }
+
+WhereClauseFact *Parser::ParseWhereClauseFact() {
+  SourceLocation Loc = Tok.getLocation();
+
+  // TODO: Handle bounds expression surrounded by parentheses, like:
+  // _Where ((((p : bounds(p, p + 1))))).
+  // Equality expressions surrounded by parentheses are already handled by
+  // invoking IgnoreValuePreservingCasts in ActOnEqualityOpFact.
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    IdentifierInfo *VarName = Tok.getIdentifierInfo();
+
+    // Consume the identifier.
+    ConsumeToken();
+    // Consume the ':' token.
+    ConsumeToken();
+
+    // Parse bounds expression.
+    ExprResult BoundsRes(ParseBoundsExpression());
+    if (BoundsRes.isInvalid())
+      return nullptr;
+    return Actions.ActOnBoundsFact(VarName, BoundsRes.get(),
+                                   getCurScope(), Loc);
+  }
+
+  ExprResult ExprRes = Actions.CorrectDelayedTyposInExpr(ParseExpression());
+  if (ExprRes.isInvalid())
+    return nullptr;
+  return Actions.ActOnEqualityOpFact(ExprRes.get(), Loc);
+}
+
+WhereClause *Parser::ParseWhereClause() {
+  // Consume the "_Where" token.
+  if (ExpectAndConsume(tok::kw__Where)) {
+    EmitDiag(Tok);
+    return nullptr;
+  }
+
+  WhereClause *WClause = Actions.ActOnWhereClause(Tok.getLocation());
+  if (!WClause) {
+    EmitDiag(Tok);
+    return nullptr;
+  }
+
+  while (true) {
+    WhereClauseFact *Fact = ParseWhereClauseFact();
+    if (!Fact) {
+      EmitDiag(Tok);
```

Comment:
It would be better to skip until an _And clause or a semicolon, whichever comes first.   Programmer time is valuable.  You want to give as many error message as possible.  You only want to bail out when you've hit a semicolon.